### PR TITLE
Support extent in directory

### PR DESCRIPTION
**Pull request by Billy4195**
Fix 2 bugs
* Fail to support longer filename
* Directory will be full if more than 128 files
Also updated the corresponding tests

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -9,7 +9,7 @@ all: $(MKFS)
 	make -C $(KDIR) M=$(PWD) modules
 
 IMAGE ?= test.img
-IMAGESIZE ?= 50
+IMAGESIZE ?= 200
```
Comment (jserv) [inline Makefile:12]: Why did you change?

Comment (jserv) [reply Makefile:12]: You shall mention the motivation along with the formula in git commit message(s) as well.

File: dir.c [hunk 2]
```diff
@@ -41,22 +43,49 @@ static int simplefs_iterate(struct file *dir, struct dir_context *ctx)
     bh = sb_bread(sb, ci->dir_block);
     if (!bh)
         return -EIO;
-    dblock = (struct simplefs_dir_block *) bh->b_data;
+    eblock = (struct simplefs_file_ei_block *) bh->b_data;
+
+    ei = (ctx->pos - 2) / SIMPLEFS_FILES_PER_EXT;
+    bi = (ctx->pos - 2) % SIMPLEFS_FILES_PER_EXT \
```
Comment (jserv) [inline dir.c]: Don't use `\` here.

File: README.md [hunk 3]
```diff
@@ -144,9 +151,9 @@ struct simplefs_extent
 ## TODO
 
 - Bugs
```
Comment (jserv) [inline README.md:153]: Don't change the items in the pull request. Instead, you can send another to improve markdown usage.

File: dir.c [hunk 4]
```diff
@@ -41,22 +43,49 @@ static int simplefs_iterate(struct file *dir, struct dir_context *ctx)
...
-            break;
-        if (!dir_emit(ctx, f->filename, SIMPLEFS_FILENAME_LEN, f->inode,
-                      DT_UNKNOWN))
+    for (; ei < SIMPLEFS_MAX_EXTENTS; ei++) {
+        if (eblock->extents[ei].ee_start == 0) {
             break;
-        ctx->pos++;
+        }
+
+        for (; bi < eblock->extents[ei].ee_len; bi++) {
```
Comment (jserv) [inline dir.c:59]: Add some comments.

File: dir.c [hunk 5]
```diff
@@ -41,22 +43,49 @@ static int simplefs_iterate(struct file *dir, struct dir_context *ctx)
...
+                    }
+                }
+                ctx->pos++;
+            }
+            brelse(bh2);
+            bh2 = NULL;
+        }
     }
 
+relse_bh:
```
Comment (jserv) [inline dir.c]: What is `relse`?

File: inode.c [hunk 6]
```diff
@@ -309,6 +368,82 @@ static int simplefs_create(struct inode *dir,
     return ret;
 }
 
+static int simplefs_remove_from_dir(struct inode *dir, struct dentry *dentry)
+{
+    struct super_block *sb = dir->i_sb;
+    //struct simplefs_sb_info *sbi = SIMPLEFS_SB(sb);
```
Comment (jserv) [inline inode.c]: Don't leave the dead code.

**Comment by jserv** [comment]
@Billy4195, do you have a chance to revise the above review messages?

**Comment by Billy4195** [comment]
@jserv OK, I'll revise my change by next weekend.
 I may not have enough time to update the README.md by the more efficient way, would it be acceptable for now?

**Comment by jserv** [comment]
> I may not have enough time to update the README.md by the more efficient way, would it be acceptable for now?

This pull request is intended to resolve extent issues. You can concentrate on the descriptions of them.

File: simplefs.h [hunk 7]
```diff
@@ -93,14 +96,17 @@ struct simplefs_extent {
 };
 
 struct simplefs_file_ei_block {
+    uint32_t nr_files; // Number of files in directory
```
Comment (jserv) [inline simplefs.h]: Always use C-style comments. That is, `/* */`

File: inode.c [hunk 8]
```diff
@@ -295,13 +317,42 @@ static int simplefs_create(struct inode *dir,
...
+        }
+        eblock->extents[ei].ee_start = bno;
+        eblock->extents[ei].ee_len = 8;
+        eblock->extents[ei].ee_block =
+            ei ? eblock->extents[ei - 1].ee_block +
+                     eblock->extents[ei - 1].ee_len
+               : 0;
+        alloc = true;
+    }
+    bh2 = NULL;
```
Comment (jserv) [inline inode.c]: This assignment is redundant.

File: inode.c [hunk 9]
```diff
@@ -316,6 +367,14 @@ static int simplefs_create(struct inode *dir,
 
     return 0;
 
+put_block:
+    if (alloc) {
```
Comment (jserv) [inline inode.c]: Rewrite as following:
```c
if (alloc && eblock->extents[ei].ee_start) {
...
}
```

File: inode.c [hunk 10]
```diff
@@ -325,6 +384,81 @@ static int simplefs_create(struct inode *dir,
...
+        }
+    }
+    if (found) {
+        if (bh_prev) {
+            brelse(bh_prev);
+        }
+        eblock->nr_files--;
+        mark_buffer_dirty(bh);
+    }
+relse_bh:
```
Comment (jserv) [inline inode.c]: relse?

File: inode.c [hunk 11]
```diff
@@ -609,35 +751,72 @@ static int simplefs_link(struct dentry *old_dentry,
...
     mark_buffer_dirty(bh);
+    brelse(bh2);
+    brelse(bh);
 
     inode_inc_link_count(inode);
     d_instantiate(dentry, inode);
+    return ret;
+
+put_block:
+    if (alloc) {
```
Comment (jserv) [inline inode.c]: Ditto.

File: inode.c [hunk 12]
```diff
@@ -609,35 +748,70 @@ static int simplefs_link(struct dentry *old_dentry,
...
     }
 
-    for (i = 0; i < SIMPLEFS_MAX_SUBFILES; i++) {
-        if (dir_block->files[i].inode == 0) {
-            f_pos = i;
-            break;
+    ei = eblock->nr_files / SIMPLEFS_FILES_PER_EXT;
+    bi = eblock->nr_files % SIMPLEFS_FILES_PER_EXT
+         / SIMPLEFS_FILES_PER_BLOCK;
+    fi = eblock->nr_files % SIMPLEFS_FILES_PER_EXT
```
Comment (jserv) [inline inode.c]: Can you avoid the computation by re-using the value of `bi` for `fi`?

File: simplefs.h [hunk 13]
```diff
@@ -8,13 +8,22 @@
...
+#define SIMPLEFS_FILENAME_LEN 255
+
+#define SIMPLEFS_FILES_PER_BLOCK \
+    (SIMPLEFS_BLOCK_SIZE / sizeof(struct simplefs_file))
+#define SIMPLEFS_FILES_PER_EXT \
+    (SIMPLEFS_FILES_PER_BLOCK *SIMPLEFS_MAX_BLOCKS_PER_EXTENT)
+
+#define SIMPLEFS_MAX_SUBFILES \
+    (SIMPLEFS_FILES_PER_EXT *SIMPLEFS_MAX_EXTENTS)
+
```
Comment (jserv) [inline simplefs.h:26]: Don't leave this blank line here.

File: inode.c (hunk 12 above)
Comment (Billy4195) [reply inode.c]: I'm not quite understand what you mean, I guess you wish to reduce the usage of modulo.
Do you mean replacing `fi` by 
`fi = eblock->nr_files - ei * SIMPLEFS_FILES_PER_EXT - bi * SIMPLEFS_FILES_PER_BLOCK`?

Comment (jserv) [reply inode.c]: What I meant was to shorten the expression.

File: dir.c [hunk 14]
```diff
@@ -38,25 +40,53 @@ static int simplefs_iterate(struct file *dir, struct dir_context *ctx)
...
+                goto release_bh;
+            }
+            dblock = (struct simplefs_dir_block *) bh2->b_data;
+            if (dblock->files[0].inode == 0) {
+                break;
+            }
+            /* Iterate every file in one block */
+            for (; fi < SIMPLEFS_FILES_PER_BLOCK; fi++) {
+                f = &dblock->files[fi];
+                if (f->inode) {
```
Comment (jserv) [inline dir.c]: Rewrite as following:
```c
if (f->inode && !dir_emit(ctx, f->filename, SIMPLEFS_FILENAME_LEN,
                               f->inode, DT_UNKNOWN))
    break;
```

File: inode.c (hunk 12 above)
Comment (Billy4195) [reply inode.c]: I found that `fi` could be changed to `fi = eblock->nr_files % SIMPLEFS_FILES_PER_BLOCK`

**Comment by jserv** [comment]
Thank @Billy4195 for the great work!
